lua/lspconfig/vala_ls.lua

```diff
@@ -5,9 +5,25 @@ configs.vala_ls = {
   default_config = {
     cmd = {'vala-language-server'},
     filetypes = {'vala', 'genie'},
-    root_dir = function(fname)
-      return util.root_pattern("meson.build")(fname)
-        or util.find_git_ancestor(fname)
+    root_dir = function (fname)
+      local cwd  = vim.loop.cwd();
+      local root = util.search_ancestors(fname, function (path)
+        local pattern = "meson.build"
+        local f = vim.fn.glob(util.path.join(path, pattern))
+        if util.path.exists(f) then
+          for line in io.lines(f) do
+            local str = line:gsub('%s+', '')
+            if str ~= '' and str:match('^project%s*%(')  then
+              return path
+            else
+              break
+            end
+          end
+        end
+      end)
+      -- prefer cwd if root is a descendant, fallback to git repo
+      return (util.path.is_descendant(cwd, root) and cwd or root) or
```

Comment:
Why prefer cwd over `util.path.dirname(fname)`?

Reply:
Removed and force pushed, please check again